### ISSUE TITLE
Generate Trie for switches on strings

### ISSUE THREAD

**Issue by YairHalberstadt** [dotnet/roslyn#43869]
See original discussion at https://github.com/dotnet/csharplang/issues/1881#issuecomment-424145818

@CyrusNajmabadi did dome testing to compare performance of current switch code gen, and a trie, and the trie was far faster.

I've repeated his tests on varying numbers of cases in the switch, and varying hit rates. The trie averages at somewhere between 2 and 4 times faster.

The only case where it was worse, was with a hundred percent hit rate, and a very small trie (containing only two items). Even then, it was just 10% worse.

I think that the downside of this is the generated IL is much larger.

If you want to play around with this, I've modified Cyrus's benchmark so that it's ready to run: https://gist.github.com/YairHalberstadt/d536fec72dcb02861ad376f983cf11c0

Copy it into a console project, run it once to generate the switches, and then uncomment the line `//#define run` to run the benchmark.

**Comment by jinujoseph** [comment]
cc @sharwell @gafter

**Comment by YairHalberstadt** [comment]
I have a branch where I have an initial implementation. I'm still liking to see if I can optimize it a bit further.

**Comment by YairHalberstadt** [comment]
I had an implementation (https://github.com/YairHalberstadt/roslyn/tree/trie-switch-strings), but when testing it I discovered there were scenarios where it was significantly slower. See https://gist.github.com/YairHalberstadt/198e42eeb7b7a4aa42f076ce5aea237a for an example (this generates c# and compiles it using the current compiler rather than using my branch, but the numbers are much the same).

I tried improving performance by using SequenceEqual in certain cases, but it turned out to be orders of magnitude slower: https://gist.github.com/YairHalberstadt/4b759e428964962e97e25d2ef681de0e. I'm not sure why, since it was far quicker in smaller benchmarks.

Either way, it's not clear there's a solid enough performance gain to be had here, so I'm leaving this for now. If somebody can come up with a better codegen strategy, I'd be happy to continue hacking on this.
